### PR TITLE
Backport e03ca73dc122af84d4a5456120e5cf5fac7aed31

### PR DESCRIPTION
**Pull request by sendaoYan**
Hi all,
This is backport of [JDK-8250825](https://bugs.openjdk.org/browse/JDK-8250825).
Patch does not apply cleanly due to different context, before [JDK-8250825](https://bugs.openjdk.org/browse/JDK-8250825)  the  PR [JDK-8230505](https://bugs.openjdk.org/browse/JDK-8230505) change these lines, and I think [JDK-8230505](https://bugs.openjdk.org/browse/JDK-8230505) do not needed backport to jdk8u-dev.

New test fails without the patch, passes with it.

Additional testing:

- [] linux x64 tier1/2/3 jtreg test
- [] linux aarch64 tier1/2/3 jtreg test

### CODE REVIEW

**Review by phohensee — APPROVED** [review]

**Review by phohensee — COMMENTED** [review]
The new test fails GHA.

**Comment by sendaoYan** [comment]
> The new test fails GHA.

Yes, that's the reason this PR still `Draft` status, I will fix the test these days.

**Comment by sendaoYan** [comment]
After the additional test finish, I will make approval request.

**Comment by sendaoYan** [comment]
> The new test fails GHA.

The test fails has been fixed now.

**Comment by sendaoYan** [comment]
The GHA report several failures:

1. macos x64 `runtime/7158988/FieldMonitor.java` fails `VM initialization failed for: j2sdk-image/jre/bin/java -Xshare:off -XX:+PrintGC -Xdebug -Xrunjdwp:transport=dt_socket,address=Mac-1721876641775.local:49199,suspend=y TestPostFieldModification`. This test also fails in [PR551](https://github.com/openjdk/jdk8u-dev/pull/551), I think it's unrelated to this PR.
2. macos x64 `compiler/unsafe/OpaqueAccesses.java` fails `SIGSEGV  in Unsafe_GetInt`, This test also fails in [PR551](https://github.com/openjdk/jdk8u-dev/pull/551), I think it's unrelated to this PR.
3. macos x64 all the `com/sun/jdi` testcases fails `transport error 202: gethostbyname: unknown host`, seems to be GHA environmental issue.

**Comment by sendaoYan** [comment]
/approval request Backport JDK-8250825, New test fails without the patch, passes with it. All the jtreg tests on linux x86_64 and aarch64 shows no new failure.

**Comment by zzambers** [comment]
> The GHA report several failures:
> 
>     1. macos x64 `runtime/7158988/FieldMonitor.java` fails `VM initialization failed for: j2sdk-image/jre/bin/java -Xshare:off -XX:+PrintGC -Xdebug -Xrunjdwp:transport=dt_socket,address=Mac-1721876641775.local:49199,suspend=y TestPostFieldModification`. This test also fails in [PR551](https://github.com/openjdk/jdk8u-dev/pull/551), I think it's unrelated to this PR.
> 
>     2. macos x64 `compiler/unsafe/OpaqueAccesses.java` fails `SIGSEGV  in Unsafe_GetInt`, This test also fails in [PR551](https://github.com/openjdk/jdk8u-dev/pull/551), I think it's unrelated to this PR.
> 
>     3. macos x64 all the `com/sun/jdi` testcases fails `transport error 202: gethostbyname: unknown host`, seems to be GHA environmental issue.

All these mac issues should be unrelated, see: https://github.com/openjdk/jdk8u-dev/pull/544

**Review by phohensee — APPROVED** [review]

**Comment by sendaoYan** [comment]
Thanks for the review.

**Comment by zzambers** [comment]
> > After #544 has been merged, does the two jbs issue can be cloesd? @zzambers
> > https://bugs.openjdk.org/browse/JDK-8337185
> > https://bugs.openjdk.org/browse/JDK-8337183

I have classified problems on MacOS in this coment: https://github.com/openjdk/jdk8u-dev/pull/544#issuecomment-2250636257

**Comment by sendaoYan** [comment]
Hi, can anyone take look this backport to jdk8u.

**Review by phohensee — APPROVED** [review]

**Comment by sendaoYan** [comment]
Thanks for the review.

File: hotspot/src/share/vm/opto/type.cpp [hunk 1]
```diff
@@ -2555,10 +2555,14 @@ TypeOopPtr::TypeOopPtr(TYPES t, PTR ptr, ciKlass* k, bool xk, ciObject* o, int o
           assert(o != NULL, "must be constant");
           ciInstanceKlass* k = o->as_instance()->java_lang_Class_klass()->as_instance_klass();
           ciField* field = k->get_field_by_offset(_offset, true);
-          assert(field != NULL, "missing field");
-          BasicType basic_elem_type = field->layout_type();
-          _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
+          if (field != NULL) {
+            BasicType basic_elem_type = field->layout_type();
+            _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
                                                        basic_elem_type == T_ARRAY);
```
Comment (gnu-andrew) [inline hotspot/src/share/vm/opto/type.cpp:2561]: This line hasn't been adjusted to line up with the one above.

Are you not backporting from the 11u patch? That also doesn't have JDK-8230505 and would match your patch exactly if not for this misaligned line.

This is the 11u version:

~~~
-          assert(field != NULL, "missing field");
-          BasicType basic_elem_type = field->layout_type();
-          _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
-                                                       basic_elem_type == T_ARRAY);
+          if (field != NULL) {
+            BasicType basic_elem_type = field->layout_type();
+            _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
+                                                         basic_elem_type == T_ARRAY);
+          } else {
+            // unsafe access
+            _is_ptr_to_narrowoop = UseCompressedOops;
+          }
~~~

**Review by gnu-andrew — COMMENTED** [review]
The change to the test for `Unsafe` looks fine and matches other 8u tests that use `Unsafe`.

I'm not sure what patch you're backporting from, but the 11u version is already adjusted for the absence of JDK-8230505. If you fix up the indentation to match 11u, this should be good to go.

File: hotspot/src/share/vm/opto/type.cpp (hunk 1 above)
Comment (sendaoYan) [reply hotspot/src/share/vm/opto/type.cpp:2561]: No, I was backported from jdk trunk repo. But I think line 2561 is not misaligned line, it just a github code display style issue. The below screenshot can explain the difference. The left side shows this PR change, the right side show the jdk11u-dev change of JDK-8250825. I think the change is equivalent.

image

Comment (sendaoYan) [reply hotspot/src/share/vm/opto/type.cpp:2561]: Backport from jdk11u-dev to jdk8u-dev is more reasonable and exactly than from jdk trunk. I will pay attention for that.

Comment (sendaoYan) [reply hotspot/src/share/vm/opto/type.cpp:2561]: I think the change is equivalent, I will resolve conversation. Backport from jdk11u-dev to jdk8u-dev is more reasonable and exactly than from jdk trunk. I will pay attention for that next time.

**Comment by sendaoYan** [comment]
/approval request Fixes the corner case in C2. Patch does not apply cleanly due to different context. New test fails without the patch, passes with it. jtreg tier{1,2,3} also passes.

**Comment by sendaoYan** [comment]
/approval request Fixes the corner case in C2. Patch does not apply cleanly due to different context. New test fails without the patch, passes with it. jtreg tier{1,2,3,etc.} also passes on linux x86_64 and linux aarch64.

**Comment by sendaoYan** [comment]
Thanks for the approved.

/integrate

**Comment by sendaoYan** [comment]
Thanks for the sponsor.

**Comment by gnu-andrew** [comment]
Why has this been pushed when the whitespace issue I raised has still not been resolved?

~~~
$ diff -u ../jdk11u-dev/src/hotspot/share/opto/type.cpp hotspot/src/share/vm/opto/type.cpp|grep basic_elem
             BasicType basic_elem_type = field->layout_type();
             _is_ptr_to_narrowoop = UseCompressedOops && (basic_elem_type == T_OBJECT ||
-                                                         basic_elem_type == T_ARRAY);
+                                                       basic_elem_type == T_ARRAY);
~~~

**Comment by gnu-andrew** [comment]
@sendaoYan I have opened [JDK-8349166](https://bugs.openjdk.org/browse/JDK-8349166) and assigned it to you. Please correct this issue.

**Comment by sendaoYan** [comment]
> @sendaoYan I have opened [JDK-8349166](https://bugs.openjdk.org/browse/JDK-8349166) and assigned it to you. Please correct this issue.

Thanks for the patient reviews. I will look the indentation issue as soon as possiable.
